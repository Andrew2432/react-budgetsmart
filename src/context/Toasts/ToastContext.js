import { createContext } from 'react';

// Create a context for toasts and export it
const ToastContext = createContext();

export default ToastContext;
