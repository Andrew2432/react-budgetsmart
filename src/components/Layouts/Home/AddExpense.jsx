import React, { useState, useContext, Fragment, useEffect } from 'react';
import ExpenseContext from '../../../context/Expense/expenseContext';
import ToastContext from '../../../context/Toasts/ToastContext';

import { CardPanel, Row, Col, TextInput } from 'react-materialize';
import BtnAddExpense from '../Buttons/BtnAddExpense';
import BtnUpdateExpense from '../Buttons/BtnUpdateExpense';
import BtnBack from '../Buttons/BtnBack';

const AddExpense = () => {
  const expenseContext = useContext(ExpenseContext);
  const toastContext = useContext(ToastContext);

  const {
    addExpense,
    mode,
    currentExpense,
    updateExpense,
    setHomeState,
    fetchDataFromStorage,
  } = expenseContext;

  const { setToast } = toastContext;

  const [expense, setExpense] = useState('');
  const [cost, setCost] = useState('');

  const handleExpense = (e) => setExpense(e.target.value);
  const handleCost = (e) => setCost(e.target.value);

  useEffect(() => {
    if (mode === 'edit') setFields();
    // eslint-disable-next-line
  }, [mode]);

  useEffect(() => {
    fetchDataFromStorage();
    // eslint-disable-next-line
  }, []);

  const handleSubmit = (e) => {
    e.preventDefault();
    if (expense.trim() === '' || !parseInt(cost)) {
      setToast({ type: 'error', text: 'Please enter correct expense.' });
    } else {
      addExpense(expense, parseInt(cost));
      clearFields();
      setToast({ type: 'success', text: 'Added successfully' });
    }
  };

  const handleUpdate = () => {
    updateExpense(expense, parseInt(cost));
    clearFields();
  };

  const handleBack = () => {
    clearFields();
    setHomeState();
  };

  const setFields = () => {
    setExpense(currentExpense.name);
    setCost(currentExpense.cost.toString());
  };

  const clearFields = () => {
    setExpense('');
    setCost('');
  };

  return (
    <CardPanel className="deep-purple white-text">
      <h4>Add A New Expense</h4>
      <Row>
        <Col s={8} m={8} l={8}>
          <TextInput
            id="expense"
            label="Expense"
            s={12}
            m={12}
            l={12}
            value={expense}
            onChange={handleExpense}
          />
        </Col>
        <Col s={4} m={4} l={4}>
          <TextInput
            id="cost"
            label="Cost"
            s={12}
            m={12}
            l={12}
            value={cost}
            onChange={handleCost}
          />
        </Col>
      </Row>
      <Row>
        {mode === 'add' && <BtnAddExpense onClick={handleSubmit} />}
        {mode === 'edit' && (
          <Fragment>
            <BtnUpdateExpense onClick={handleUpdate} />
            <BtnBack onClick={handleBack} />
          </Fragment>
        )}
      </Row>
    </CardPanel>
  );
};

export default AddExpense;
