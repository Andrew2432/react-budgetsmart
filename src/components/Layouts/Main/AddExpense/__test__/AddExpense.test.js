import TestRenderer from 'react-test-renderer';
import React from 'react';
import AddExpense from '../AddExpense';
import ToastContext from '../../../../../context/Toasts/ToastContext';
import ExpenseContext from '../../../../../context/Expense/expenseContext';

describe('AddExpense component', () => {
  test('Renders correctly', () => {
    const addExpense = new TestRenderer.create(
      (
        <ExpenseContext.Provider value={{ expenses: [] }}>
          <ToastContext.Provider value={{ messages: [] }}>
            <AddExpense />
          </ToastContext.Provider>
        </ExpenseContext.Provider>
      )
    );
    expect(addExpense).toMatchSnapshot();
  });
});
