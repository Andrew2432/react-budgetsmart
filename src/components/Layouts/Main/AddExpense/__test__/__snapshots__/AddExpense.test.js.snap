// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddExpense component Renders correctly 1`] = `
<div
  className="card-panel deep-purple white-text"
>
  <h4>
    Add A New Expense
  </h4>
  <div
    className="row"
  >
    <div
      className="col s8 m8 l8"
    >
      <div
        className="input-field col s12 m12 l12"
      >
        <input
          className=""
          id="expense"
          onChange={[Function]}
          type="text"
          value=""
        />
        <label
          className=""
          htmlFor="expense"
        >
          Expense
        </label>
      </div>
    </div>
    <div
      className="col s4 m4 l4"
    >
      <div
        className="input-field col s12 m12 l12"
      >
        <input
          className=""
          id="cost"
          onChange={[Function]}
          type="text"
          value=""
        />
        <label
          className=""
          htmlFor="cost"
        >
          Cost
        </label>
      </div>
    </div>
  </div>
  <div
    className="row"
  />
</div>
`;
